src/main/java/net/glowstone/util/mojangson/Mojangson.java

```diff
@@ -0,0 +1,542 @@
+                    }
+                    Tag val = (Tag) parseTag(tmpval);
+
+                    if (tagType == null) {
+                        tagType = val.getType();
+                    } else if (tagType != val.getType()) {
+                        throw new MojangsonParseException("Index: " + index + ", value: \'" + tmpval + "\'", MojangsonParseException.ParseExceptionReason.INCOMPATIBLE_TYPE);
+                    }
+
+                    values.add(val);
+                    tmpval = "";
+                    continue;
+                }
+                tmpval += character;
+            }
+        }
+        if (tagType == TagType.BYTE) {
+            byte[] bytes = new byte[values.size()];
+            for (int i = 0; i < values.size(); i++) {
+                bytes[i] = (byte) values.get(i).getValue();
+            }
+            return new ByteArrayTag(bytes);
+        } else if (tagType == TagType.INT) {
+            int[] ints = new int[values.size()];
+            for (int i = 0; i < values.size(); i++) {
+                ints[i] = (int) values.get(i).getValue();
+            }
+            return new IntArrayTag(ints);
+        } else {
+            return new ListTag<>(tagType, values);
+        }
+    }
+
+    /**
+     * Creates a Mojangson string from the given NBT Tag. Convenience method for other fromX(Tag) methods.
+     *
+     * @param tag the NBT Tag to convert
+     * @return the converted Mojangson string
+     */
+    public static String fromTag(Tag tag) {
+        if (tag.getType() == TagType.BYTE) {
+            return fromByteTag((ByteTag) tag);
+        } else if (tag.getType() == TagType.BYTE_ARRAY) {
+            return fromByteArrayTag((ByteArrayTag) tag);
+        } else if (tag.getType() == TagType.COMPOUND) {
+            return fromCompoundTag((CompoundTag) tag);
+        } else if (tag.getType() == TagType.DOUBLE) {
+            return fromDoubleTag((DoubleTag) tag);
+        } else if (tag.getType() == TagType.FLOAT) {
+            return fromFloatTag((FloatTag) tag);
+        } else if (tag.getType() == TagType.INT) {
+            return fromIntTag((IntTag) tag);
+        } else if (tag.getType() == TagType.INT_ARRAY) {
+            return fromIntArrayTag((IntArrayTag) tag);
+        } else if (tag.getType() == TagType.LIST) {
+            return fromListTag((ListTag) tag);
+        } else if (tag.getType() == TagType.LONG) {
+            return fromLongTag((LongTag) tag);
+        } else if (tag.getType() == TagType.SHORT) {
+            return fromShortTag((ShortTag) tag);
+        } else if (tag.getType() == TagType.STRING) {
+            return fromStringTag((StringTag) tag);
+        }
+        return String.valueOf(tag.getValue());
+    }
+
+    /**
+     * Creates a Mojangson string from the given Byte Tag.
+     *
+     * @param tag the Byte Tag to convert
+     * @return the converted Mojangson string
+     */
+    public static String fromByteTag(ByteTag tag) {
```

Comment:
I think these should still be named fromTag, but have different parameter types, and then you can just remove the huge tree for the method above.